Add is-popup to layout div when opening login form in a popup

gravityrail opened the pull request:
We want to add this class so we can re-style the connection when it's open during the wp-admin connect flow.

#### Changes proposed in this Pull Request

* Add `is-popup` to the `layout` div when showing the login or create account screens during Jetpack's "wp-admin" connect flow.

Example div classes: `layout is-section-login focus-content has-no-sidebar has-no-masterbar is-jetpack-login is-popup`

#### Testing instructions

* Got through the wp-admin connect flow from a recent Jetpack master build ( you will need to set `define( 'JETPACK_SHOULD_USE_CONNECTION_IFRAME', true );` in `wp-config.php`)
* When the popup opens, you should see `is-popup` as a class on the `layout` div inside the `wpcom` div.

* Setting `calypso_env` for the wp-admin connect flow is part of a separate patch (ping @gravityrail )

gravityrail committed:
Add is-popup to layout div when opening login form in a popup

jsnajdr commented on client/layout/logged-out.jsx:
```diff
@@ -134,6 +136,7 @@ export default connect( state => {
 	const currentRoute = getCurrentRoute( state );
 	const isJetpackLogin = currentRoute === '/log-in/jetpack';
 	const noMasterbarForRoute = currentRoute === '/log-in/jetpack';
+	const isPopup = typeof window !== 'undefined' && window.opener && window.opener !== window;
```
Ideally, server-rendered markup should be identical to the client-rendered one. React is rather sensitive about that and mail fail to hydrate the SSR-ed page.

Here, the SSR-ed page requested by the popup window will be rendered with `isPopup=false`, and the initial client-side render will be rendered with `isPopup=true`. Not 100% isomorphic.

Adding a `popup=true` query arg would be one of possible solutions.

roccotripaldi replied:
Thanks @jsnajdr - working on that server-side in D33024-code

roccotripaldi committed:
Jetpack Connect: adding `is_popup` to authorize query schema

Certain connection flows maybe opened within a pop-up window,
and we detect that by looking at URL params.

Adding this property to the authorize query schema will allow the
property to persists if the person clicks through different routes
during the connection process.

roccotripaldi committed:
Layout classes

If `is_popup` is present in the URL, we add a similar class to main layout classes.
This will allow us to apply different styling and logic to popuped flows.

roccotripaldi committed:
updating jest snap shot

gravityrail commented on client/jetpack-connect/schema.js:
```diff
@@ -24,6 +24,7 @@ export const authorizeQueryDataSchema = {
 		close_window_after_login: { type: 'string' },
 		from: { type: 'string' },
 		home_url: { type: 'string' },
+		is_popup: { type: 'string' },
```
How does this behave with various truth/falsy values in the URL? I see you are interpreting the value as a string, which could cause unintended consequences.

<img width="117" alt="negating-url-params" src="https://user-images.githubusercontent.com/51896/65463039-d7521c80-de0b-11e9-877b-6a161432ce9f.png">

roccotripaldi replied:
good call. I made this more explicit in 11c743240ea18721e466b1931ccb72bf9373c965
and included a unit test for good measure.

gravityrail reviewed (commented):
@roccotripaldi

roccotripaldi committed:
Jetpack connect: coerce truthy strings to true

Data coming through URL params will always be treated as a string.
When we transform data, let's coerce truthy strings to booleans.

simison commented on client/jetpack-connect/utils.js:
```diff
@@ -14,11 +14,13 @@ import { head, includes, isEmpty, split } from 'lodash';
-		closeWindowAfterLogin: !! queryObject.close_window_after_login,
+		closeWindowAfterLogin: -1 !== truthy.indexOf( queryObject.close_window_after_login ),
 		homeUrl: queryObject.home_url,
+		isPopup: -1 !== truthy.indexOf( queryObject.is_popup ),
```
You could use ES6 version instead: `truthy.includes( queryObject.is_popup )` but I'm curious why we can't rely URL argument always be the same?

roccotripaldi replied:
good call. yeah, i suppose we can enforce it to always be 1

roccotripaldi replied:
addressd in 5bc0e3e017f3138aac9245845ef4548c678eba40

roccotripaldi committed:
Jetpack connect

- enforce `is_popup` and `close_window_after_login` to be '1' if true.

enejb commented:
I tested this and I wasn't able to see the is-popup class. 
I was able to see `is_popup=1` in the url but it was encoded. 

Also it seems that the PR needs a rebase.

roccotripaldi commented:
Thanks for the review @enejb - I rebased.
I also noticed that when viewing the page's source, i don't see an `is-popup` class on the layout div. I think that's a quirk of browser tools and react not always playing nicely together.

Because when i add a rule like:

```
.layout.is-popup {
	color: red !important;
}
```

i end up with this:

<img width="612" alt="Screen Shot 2019-09-26 at 1 36 25 PM" src="https://user-images.githubusercontent.com/2694219/65711469-ef6fa900-e062-11e9-8dd9-c1126feeb4da.png">

So i think we are golden gravy, m'dude!

enejb commented:
Nice work Rocco! :shipit:

jsnajdr reviewed (approved):
The code is good and it also works well: `is_popup=1` query parameter leads to a `is-popup` CSS class on the layout element. It works only in logged-out layout, though. I assume that's exactly what you wanted.
